include/openthread/message.h

```diff
@@ -69,6 +69,19 @@ typedef enum otMessagePriority
     OT_MESSAGE_PRIORITY_HIGH   = 2, ///< High priority level.
 } otMessagePriority;
 
+/**
+ * Defines the OpenThread message origins.
+ *
+ */
+typedef enum otMessageOrigin
+{
+    OT_MESSAGE_ORIGIN_FROM_THREAD_NETIF           = 0, ///< Message originates from Thread Netif.
+    OT_MESSAGE_ORIGIN_FROM_HOST_DISALLOW_LOOPBACK = 1, ///< Message originates from host and should not be passed back
+                                                       ///< to host.
+    OT_MESSAGE_ORIGIN_FROM_HOST_ALLOW_LOOPBACK = 2,    ///< Message originates from host and can be passed back to host.
+    OT_MESSAGE_ORIGIN_FROM_HOST_UNTRUSTED      = 3,    ///< Message originates from an untrusted source on host.
+} otMessageOrigin;
+
```

Comment:
How about we instead add APIs to allow caller to set the origin on an allocated `otMessage`. 

Then we can still use the `otIp6Send()` and don't need to add a new API?

We can make sure when we allocate an `otMessage *` for host (from `otIp6NewMessage()` or `otUdpNewMessage()`, ...)  we set its origin such that current behavior remains unchanged:
- Today in `otIp6Send()` which calls `Ip6::SendRaw()` we determine the origin using `OPENTHREAD_CONFIG_IP6_ALLOW_LOOP_BACK_HOST_DATAGRAMS`. 
- We can use the same logic to set the origin on a newly allocated message from `otIp6NewMessage()` or `otUdpNewMessage()`, ...).
- The user can then change it to `OT_MESSAGE_ORIGIN_FROM_HOST_UNTRUSTED` if he/she wants.

Reply:
Thanks. I implemented this idea in the latest push.

Reply:
@abtink Do we really want to expose `OT_MESSAGE_ORIGIN_THREAD_NETIF`, `OT_MESSAGE_ORIGIN_HOST_DISALLOW_LOOPBACK` and `OT_MESSAGE_ORIGIN_HOST_ALLOW_LOOPBACK` as public APIs? This is going to allow applications to set whatever origin they want and fake Thread packets.

Should we always set OT_MESSAGE_ORIGIN_FROM_HOST_UNTRUSTED for messages with `otIp6Send`?

Reply:
> Do we really want to expose `OT_MESSAGE_ORIGIN_THREAD_NETIF`, `OT_MESSAGE_ORIGIN_HOST_DISALLOW_LOOPBACK` and `OT_MESSAGE_ORIGIN_HOST_ALLOW_LOOPBACK` as public APIs? This is going to allow applications to set whatever origin they want and fake Thread packets.

I think it is fine to define the full list of `OT_MESSAGE_ORIGIN` values. 

We can provide `otMessageGetOrigin()` as well. This way the user (host or next layer) can use it to learn the origin of a message we are passing back to host (in the callbacks) and add any other filtering they want.

Regarding what user can use:
- I think user should be able to choose to use `ORIGIN_HOST_DISALLOW_LOOPBACK` or `HOST_ALLOW_LOOPBACK` or `HOST_UNTRUSTED` if they want to control behavior they want for the message.
-  But we should disallow `ORIGIN_THREAD`. We can add check for thisand return error from `otIp6Send()` API or other APIs that get a message from host/user if `ORIGIN_THREAD` is set on the given `otMessage`. Thoughts?

----------

> Should we always set `OT_MESSAGE_ORIGIN_FROM_HOST_UNTRUSTED` for messages with `otIp6Send`?

I am not sure.  We have to be careful that the `otIp6Send()` is an existing API that is being used not just on BR but all types of devices (on end-device, routers, sleepy-devices, etc).

Anytime we change APIs, we should try to keep existing behaviors as they are so not to break any existing code that uses them and may rely on certain existing behavior.

So that's why I suggested this so the existing behavior remains unchanged:

> We can make sure when we allocate an `otMessage *` for host (from `otIp6NewMessage()` or `otUdpNewMessage()`, ...)  we set its origin such that current behavior remains unchanged:
> - Today in `otIp6Send()` which calls `Ip6::SendRaw()` we determine the origin using `OPENTHREAD_CONFIG_IP6_ALLOW_LOOP_BACK_HOST_DATAGRAMS`. 
> - We can use the same logic to set the origin on a newly allocated message from `otIp6NewMessage()` or `otUdpNewMessage()`, ...).
> - The user can then change it to `OT_MESSAGE_ORIGIN_FROM_HOST_UNTRUSTED` if he/she wants.
>

Reply:
> Anytime we change APIs, we should try to keep existing behaviors as they are so not to break any existing code that uses them and may rely on certain existing behavior.

+1